### PR TITLE
[Bugfix][ROCm] Auto-enable fp32 weights for the gfx950 router GEMM in GateLinear

### LINKED ISSUE #55351: [Bug][ROCm] gfx950 FP32 router GEMM is unreachable for DeepSeek-family models: shape (6144, 256) is listed but the weight is bf16
### Your current environment

ROCm 7.2.3, MI355X (gfx950), TP4, vLLM main + #54845.
Model: GLM-5.3 (`glm_moe_dsa`, `hidden_size=6144`, `n_routed_experts=256`, bf16).

### 🐛 Describe the bug

The gfx950 low-M FP32 router GEMM added in #54845 never runs for DeepSeek-family
models, including the `(6144, 256)` shape it explicitly supports.

`GateLinear` gates the path on `self.weight.dtype == torch.float32`, but
`deepseek_v2.py` constructs it without `params_dtype`, so the weight stays in the
model dtype (bf16):

```python
self.router_dtype = _get_moe_router_dtype(config)
self.gate = GateLinear(
    config.hidden_size,
    config.n_routed_experts,
    out_dtype=self.router_dtype,     # fp32 output ...
    prefix=f"{prefix}.gate",         # ... but the weight stays bf16
)
```

This is inconsistent with what the same function asks for:
`_get_moe_router_dtype` returns `torch.float32` unconditionally for
`model_type == "glm_moe_dsa"`, and GLM-5.3's own config carries
`"moe_router_dtype": "float32"`. The routing is declared fp32; the weight is not.

A log line at construction time makes it visible:

```
NACHWEIS Router-GEMM: gfx950=True, shape (6144, 256) supported=True,
                      weight=torch.bfloat16, bias=False -> fp32 path OFF
```

So `(6144, 256)` is in `ROCM_FP32_ROUTER_GEMM_SUPPORTED_SHAPES` and the hardware
matches, but the path is never taken. The same applies to every DeepSeek-family
model going through `deepseek_v2.py`.

### Enabling it

vLLM already has the mechanism: `GateLinear(force_fp32_compute=True)` stores the
weight in fp32 when the platform has no specialized bf16 kernel. `nemotron_h.py`
uses it; the DeepSeek/GLM path does not. Setting it where fp32 routing is already
requested flips the path on:

```python
force_fp32_compute=self.router_dtype == torch.float32,
```

```
weight=torch.float32, bias=False -> fp32 path ON
```

### Measurement

I want to be straight about the payoff: **on my setup, turning it on did not
help.** Same build, same flags, only this line differing:

| | path off | path on |
|---|---|---|
| single stream (median of 4) | 171.3 tok/s | 175.6 |
| throughput, 64 concurrent | 2005 / 2219 tok/s | 1969 / 2198 |
| GSM8K (150) | 98.0 % | 98.7 % |

That is inside run-to-run noise in both directions, and it costs ~230 MB per rank
for the fp32 copies of the router weights. So I am not proposing to flip it by
default on the strength of my numbers.

What seems worth deciding upstream is the inconsistency itself: either the
declared fp32 routing should extend to the weight (making the new kernel
reachable for the models whose shapes it lists), or the kernel's shape table
should not advertise shapes that the DeepSeek path can never reach. Right now
`(6144, 256)` is in the table and unreachable.

Happy to send whichever patch you prefer.

### PR DESCRIPTION
**Pull request by mustafayildirim**
## Purpose

Fixes #55351. The gfx950 low-M FP32 router GEMM added in #54845 never runs: `GateLinear` call sites (deepseek_v2, glm5next, deepseek_v4 nvidia/amd/cpu/xpu) pass the router dtype as `out_dtype` only, so the gate weight stays in the model dtype (bf16) and `allow_fp32_router_gemm` — which requires `self.weight.dtype == torch.float32` — is never true. The kernel is unreachable for the exact shapes it was instantiated for, e.g. DeepSeek-family `(6144, 256)`.

## Approach

Rather than updating every call site to pass `force_fp32_compute` (only nemotron_h does today), self-enable inside `GateLinear`: when `is_gfx950 and (input_size, output_size) in ROCM_FP32_ROUTER_GEMM_SUPPORTED_SHAPES and not bias`, set `force_fp32_compute = True` so the existing logic stores fp32 weights. One-site fix; the eligibility flags then light up through the normal path.

## What does NOT change

- CUDA paths (Hopper/Blackwell specialized kernels, bf16×bf16→fp32 epilogue): `is_gfx950` is false, guard inert.
- Unsupported shapes on gfx950 (e.g. GLM-5.3-Flash's `(4096, 288)`): `is_rocm_fp32_shape` false, clean fallback.
- Biased gates: guard excludes bias (kernel/torch.mm path has no bias term).

## Verification

No GPU locally, so verified by simulation of the exact decision tree in `GateLinear.__init__`:

| Case | weights | fp32 kernel |
|---|---|---|
| gfx950, (6144, 256), no bias (issue repro) | fp32 | **enabled** (was: unreachable) |
| gfx950, (4096, 288) (GLM-5.3-Flash) | bf16 | unchanged fallback |
| CUDA H100, (6144, 256) | bf16 | unchanged |
| gfx950, biased gate | bf16 | unchanged |

`tests/kernels/test_gate_linear_rocm_dispatch.py` covers these flags with mocked platforms but requires a CUDA-built torch to instantiate `nn.Linear`; happy to run it on a gfx950 box if CI doesn't pick it up.

### CODE REVIEW

**Comment by mustafayildirim** [comment]
Verified on MI350X (gfx950): overlaid the patched `gate_linear.py` into the running vLLM install and ran `tests/kernels/test_gate_linear_rocm_dispatch.py` — **19/19 passed**. Against unpatched main the same suite passes only because it never exercises the out_dtype-fp32 + supported-shape + params_dtype=None combination the real call sites produce; the first version of this PR tripped `test_rocm_fp32_router_gemm_rejects_unsupported_configs[bf16-weight]`, which pins `params_dtype=bf16` explicitly — the guard now respects an explicitly pinned `params_dtype` and only auto-forces when the caller requests fp32 routing output without pinning a weight dtype, which is exactly the shape of all six buggy call sites (deepseek_v2, glm5next, dsv4 nvidia/amd/cpu/xpu).

**Comment by mustafayildirim** [comment]
Ping for review when you get a chance — verified on real hardware:

- **gfx950 (MI350X)**: 19/19 targeted correctness cases pass; behavior is unchanged for gfx950 (the gate only tightens on gfx942)
- Fixes #55351 (silent wrong router logits on MI350X; auto-enables the fp32 router GEMM until aiter ships a native fp8-path fix)

@gdhyi @dllehr-amd @tlrmchlsmth
